### PR TITLE
Add RPC method "bp_addTx" and "bp_getRunningStatus", also support JSON-RPC for miners

### PR DESCRIPTION
**Pull request by ggicci**
1. add rpc method: bp_getRunningStatus;
2. add rpc method: bp_addTx;
3. support JSON RPC API for cql-minerd;
4. some refactoring;
5. add unittests for package: rpc/jsonrpc.

### CODE REVIEW

**Comment by ggicci** [comment]
Yes, i'm going to make the PR bigger and bigger ...

File: crypto/kms/pubkeystore.go [hunk 1]
```diff
@@ -233,7 +233,9 @@ func SetNode(nodeInfo *proto.Node) (err error) {
 	if nodeInfo == nil {
 		return ErrNilNode
 	}
-	if !Unittest {
+
+	// FIXME(ggicci): client node id is a fake one.
+	if !Unittest && nodeInfo.Role != proto.Client {
```
Comment (xq262144) [inline crypto/kms/pubkeystore.go:238]: Client `NodeID` and nonce still needs verification.

File: rpc/jsonrpc/registerclient.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	expectedAddr, err := crypto.PubKeyHash(pubKey)
+	if err != nil {
+		return nil, errors.WithMessage(err, "invalid public key: generate address error")
+	}
+	if expectedAddr.String() != params.Address {
+		return nil, errors.New("invalid pair of address and public key")
+	}
+
+	nodeInfo := &proto.Node{
+		ID:        proto.NodeID(params.Address), // a fake node id
```
Comment (xq262144) [inline rpc/jsonrpc/registerclient.go:49]: Should use node miner to mine a valid `NodeID`.

File: rpc/jsonrpc/websocket_test.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+			name                 string
+			method               string
+			params               interface{}
+			expectedError        error
+			expectedResult       interface{}
+			expectedResultAssert func(actual interface{}, expected ...interface{}) string
+		}{
+			{
+				name:          "nil method should not be found",
+				method:        "nil",
```
Comment (xq262144) [inline rpc/jsonrpc/websocket_test.go:124]: Wrong test case params here.

File: crypto/kms/privatekeystore.go [hunk 4]
```diff
@@ -146,11 +146,12 @@ func InitLocalKeyPair(privateKeyPath string, masterKey []byte) (err error) {
...
 	if err != nil {
 		log.WithError(err).Info("load private key failed")
-		if errors.Cause(err) == ErrNotKeyFile {
+		cause := errors.Cause(err)
+		if cause == ErrNotKeyFile {
 			log.WithField("path", privateKeyPath).Error("not a valid private key file")
 			return
 		}
-		if _, ok := err.(*os.PathError); (ok || err == os.ErrNotExist) && conf.GConf.GenerateKeyPair {
+		if os.IsNotExist(cause) && conf.GConf.GenerateKeyPair {
```
Comment (auxten) [inline crypto/kms/privatekeystore.go:154]: This is better.

**Pull request closed without merging**